04-Protocol-Security.md

```diff
@@ -331,29 +331,67 @@ If the server provides a non-empty `CIPHER_CHOICE`:
 2. New keys `key_new` are derived from the original CipherState keys `key_orig` by taking the first 32 bytes from `ENCRYPT(key_orig, maxnonce, zero_len, zeros)` using the negotiated cipher function where `maxnonce` is 2<sup>64</sup> - 1, `zerolen` is a zero-length byte sequence, and `zeros` is a sequence of 32 bytes filled with zeros. (see `Rekey(k)` function<sup>[8](#reference-8)</sup>)
 3. New CipherState objects are reinitialized: `InitializeKey(key_new)`.
 
-### 4.5.6 Transport message encryption and format
-
-After handshake process is finished, both initiator and responder have CipherState objects for encryption and decryption and after initiator validated server's identity, any subsequent traffic is encrypted and decrypted with `EncryptWithAd()` and `DecryptWithAd()` methods of the respectrive CipherState objects with zero-length associated data.
-
-Ciphertext is sent in `NOISE_FRAME` over the wire.
-
-##### NOISE_FRAME
-| Field Name | Data Type | Description |
-| ---------- | --------- | ----------- |
-| ciphertext | B0_64K | AEAD ciphertext including 16 bytes MAC |
-
-Message length: `<Plaintext length> + 18 bytes = <Plaintext length> + <MAC length> + <Type length prefix>`
-
-Maximum message length = 65537 bytes
-Maximum ciphertext length = 65535 bytes
-Maximum plaintext length 65519 bytes
+## 4.6 Encrypted stratum message framing
+
+After handshake process is finished, both initiator and responder have CipherState objects for encryption and decryption and after initiator validated server's identity, any subsequent traffic is encrypted and decrypted with `EncryptWithAd()` and `DecryptWithAd()` methods of the respective CipherState objects with zero-length associated data.
+
+Maximum transport message length (ciphertext) is for noise protocol message 65535 bytes.
+
+Since Stratum Message Frame consists of
+- fixed length message header: 6 bytes
+- variable length serialized stratum message
+
+Stratum Message header and stratum message payload are processed separately.
+
+#### Encrypting stratum message
+1. serialize stratum message into a plaintext binary string (payload)
+2. prepare frame header for the stratum message with `message_length` being length of payload ciphertext*
```

Comment:
Open question:

Should the (encrypted) Frame header contain the `message_length` of the plaintext serialized message or  that of the ciphertext?

proposed change uses ciphertext length.
using this option means the field tells "how many bytes to read next" (length is calculated by sender). The interpretation is straightforward.

If we used plaintext serialized message length, then the receiver would need to calculate how many bytes he should read. The interpretation is not as straightforward.

Reply:
As discussed on last dev call everyone agree on having the len of the ciphertext